[Spring 지하철 노선도 - 3단계] 스컬(김윤기) 미션 제출합니다.

skullkim opened the pull request.

skullkim committed:
refactor(LineDao.java): dto를 도메인 객체로 대체

컨트롤러에서 요청을 받을때 사용되는 dto를 파라미터로 받는 부분을 도메인
객체로 변경

BREAKING CAHNGE: LineRequest를 Line으로 변경
as-is:
@Repository
public class LineDao {
    ...

    public void update(final Long id, final LineRequest lineRequest) {
        ...
    }
}

to-be:
public class LineDao {
...

    public void update(final Long id, final Line line) {
        ...
    }
}

skullkim committed:
refactor(service/): 인터페이스 삭제

서비스 계층과 컨트롤러 계층의 분리를 위한 인터페이스 사용이 불필요하다
느껴 서비스 인터페이스 삭제

BREAKING CHANGE: CommonLineService, CommonStationService 삭제
as-is:
@Service
public class LineService implements CommonLineService {

   ...

    @Override
    ...
    public Line save(final LineRequest lineRequest) {
        ...
    }

    @Override
    ...
    public Line findById(final Long id) {
        ...
    }

    @Override
    ...
    public List<Line> findAll() {
        ...
    }

    @Override
    ...
    public void update(final Long id, final LineRequest lineRequest) {
        ...
    }

    @Override
    ...
    public void deleteById(final Long id) {
        ...
    }
}

@Service
public class StationService implements CommonStationService {

    ...

    ...
    @Override
    public Station save(final StationRequest stationRequest) {
        final Station station = new Station(stationRequest.getName());
        return stationDao.save(station);
    }

    ...
    @Override
    public List<Station> findAll() {
        return stationDao.findAll();
    }

    ...
    @Override
    public void deleteById(final Long id) {
        ...
    }
}

to-be:
@Service
public class LineService {

   ...

    ...
    public Line save(final LineRequest lineRequest) {
        ...
    }

    ...
    public Line findById(final Long id) {
        ...
    }

    ...
    public List<Line> findAll() {
        ...
    }

    ...
    public void update(final Long id, final LineRequest lineRequest) {
        ...
    }

    ...
    public void deleteById(final Long id) {
        ...
    }
}

@Service
public class StationService {

    ...

    ...
    public Station save(final StationRequest stationRequest) {
        final Station station = new Station(stationRequest.getName());
        return stationDao.save(station);
    }

    ...
    public List<Station> findAll() {
        return stationDao.findAll();
    }

    ...
    public void deleteById(final Long id) {
        ...
    }
}

skullkim committed:
refactor(dao/): 인터페이스 적용

서비스와 자주 변경되는 레파지토리 레이어를 분리하기 위해 인터페이스
적용

BREAKING CHANGE: dao에 인터페이스 적용
as-is:
@Repository
public class LineDao {

    public LineDao(final NamedParameterJdbcTemplate namedParameterJdbcTemplate, final DataSource dataSource) {
        ...
    }

    public Line save(final Line line) {
        ...
    }

    public List<Line> findAll() {
        ...
    }

    public Line findById(final Long id) {
        ...
    }

    public void update(final Long id, final LineRequest lineRequest) {
        ...
    }

    public void deleteById(final Long id) {
        ...
    }
}

@Repository
public class StationDao {

    ...

    public StationDao(final NamedParameterJdbcTemplate namedParameterJdbcTemplate, final DataSource dataSource) {
        ...
    }

    public Station save(final Station station) {
        ...
    }

    public List<Station> findAll() {
        ...
    }

    public void deleteById(final Long id) {
        ...
    }
}

to-be:
@Repository
public class LineDao implements CommonLineDao {

    ...

    public LineDao(final NamedParameterJdbcTemplate namedParameterJdbcTemplate, final DataSource dataSource) {
        ...
    }

    @Override
    public Line save(final Line line) {
        ...
    }

    @Override
    public List<Line> findAll() {
        ...
    }

    @Override
    public Line findById(final Long id) {
        ...
    }

    @Override
    public void update(final Long id, final Line line) {
        ...
    }

    @Override
    public void deleteById(final Long id) {
        ...
    }
}

@Repository
public class StationDao implements CommonStationDao {

    ...

    public StationDao(final NamedParameterJdbcTemplate namedParameterJdbcTemplate, final DataSource dataSource) {
        ...
    }

    @Override
    public Station save(final Station station) {
        ...
    }

    @Override
    public List<Station> findAll() {
        ...
    }

    @Override
    public void deleteById(final Long id) {
        ...
    }
}

skullkim committed:
chore(test/java/wooteco/subway/service): 테스트 클래스 이름 변경

실제 구현의 service 이름이 바뀜에 따라 테스트 클래스 이름 변경

BREAKING CHANGE: 서비스 테스트 클래스 이름 변경
as-is:
SpringLineService
SpringStationService

to-be:
LineService
StationService

skullkim committed:
refactor(LineDao.java): 예외처리 제거

dao는 영속 데이터에 접근하는 역할만 해야 하므로 예외 처리 부분을
서비스로 이동

BREAKING CAHNGE: dao에 존재하던 예외처리 로직 서비스로 이동
as-is:
@Repository
public class LineDao {

    private static final int NO_ROW_AFFECTED = 0;
    private static final String LINE_DUPLICATED = "이미 존재하는 노선입니다. ";
    private static final String LINE_NOT_FOUND = "요청한 노선이 존재하지 않습니다. ";

    ...

    public Line save(final Line line) {
        ...
        try {
            ...
        } catch (DuplicateKeyException e) {
            throw new IllegalStateException(LINE_DUPLICATED + line);
        }
    }

    ...

    public void update(final Long id, final LineRequest lineRequest) {
        ...
        if (theNumberOfAffectedRow == NO_ROW_AFFECTED) {
            throw new IllegalStateException(LINE_NOT_FOUND + "id=" + id + " " + lineRequest);
        }
    }

    public void deleteById(final Long id) {
        ...
        if (theNumberOfAffectedRow == NO_ROW_AFFECTED) {
            throw new IllegalStateException(LINE_NOT_FOUND + "id=" + id);
        }
    }
}

@Service
public class LineService implements CommonLineService {
    ...

     @Override
    @Transactional
    public Line save(final LineRequest lineRequest) {
        final Line line = new Line(lineRequest.getName(), lineRequest.getColor());
        return lineDao.save(line);
    }

    ...

    @Override
    @Transactional
    public void update(final Long id, final LineRequest lineRequest) {
        lineDao.update(id, lineRequest);
    }

    @Override
    @Transactional
    public void deleteById(final Long id) {
        lineDao.deleteById(id);
    }
}

to-be:
@Repository
public class LineDao implements CommonLineDao {
    ...

    @Override
    public Line save(final Line line) {
        ...
        final Long id = simpleInsert.executeAndReturnKey(params).longValue();
        return new Line(id, line.getName(), line.getColor());
    }

    ...

    @Override
    public int update(final Long id, final Line line) {
        ...
        ...
        return namedParameterJdbcTemplate.update(sql, parameter);
    }

    @Override
    public int deleteById(final Long id) {
        ...
        return namedParameterJdbcTemplate.update(sql, Map.of("id", id));
    }
}

@Service
public class LineService {

    private static final int NO_ROW_AFFECTED = 0;
    private static final String LINE_DUPLICATED = "이미 존재하는 노선입니다. ";
    private static final String LINE_NOT_FOUND = "요청한 노선이 존재하지 않습니다. ";

    ...

    @Transactional
    public Line save(final LineRequest lineRequest) {
        ...
        try {
            return lineDao.save(line);
        } catch (DuplicateKeyException e) {
            throw new IllegalStateException(LINE_DUPLICATED + line);
        }
    }

    ...

    @Transactional
    public void update(final Long id, final LineRequest lineRequest) {
        ...
        if (theNumberOfAffectedRow == NO_ROW_AFFECTED) {
            throw new IllegalStateException(LINE_NOT_FOUND + "id=" + id + " " + line);
        }
    }

    @Transactional
    public void deleteById(final Long id) {
        ...
        if (theNumberOfAffectedRow == NO_ROW_AFFECTED) {
            throw new IllegalStateException(LINE_NOT_FOUND + "id=" + id);
        }
    }
}

skullkim committed:
refactor(StationDao.java): 예외처리 로직 제거

dao는 영속 데이터에 접근하는 역할만 해야 하므로 예외 처리 로직 제거

BREAKING CHANGE: StatinoDao에 존재하는 예외처리 로직을 StationService로
이동
as-is:
@Repository
public class StationDao {

    private static final int NO_ROW_AFFECTED = 0;
    private static final String STATION_DUPLICATED = "이미 존재하는 지하철역입니다. ";
    private static final String STATION_NOT_FOUND = "요청한 지하철 역이 존재하지 않습니다. ";

    ...

    public Station save(final Station station) {
        ...
        try {
            ...
        } catch (DuplicateKeyException e) {
            throw new IllegalStateException(STATION_DUPLICATED + station);
        }
    }

    ...

    public void deleteById(final Long id) {
        ...
        if (theNumberOfAffectedRow == NO_ROW_AFFECTED) {
            throw new IllegalStateException(STATION_NOT_FOUND + "id=" + id);
        }
    }
}

@Service
public class StationService implements CommonStationService {
    ...

   @Transactional
    @Override
    public Station save(final StationRequest stationRequest) {
        final Station station = new Station(stationRequest.getName());
        return stationDao.save(station);
    }

    ...

    @Transactional
    @Override
    public Station save(final StationRequest stationRequest) {
        final Station station = new Station(stationRequest.getName());
        return stationDao.save(station);
    }
}

to-be:
@Repository
public class StationDao implements CommonStationDao {
    ...

    @Override
    public Station save(final Station station) {
        ...
        final Long id = simpleInsert.executeAndReturnKey(params).longValue();
        return new Station(id, station.getName());
    }

    ...

    @Override
    public int deleteById(final Long id) {
        final String sql = "delete from STATION where id = :id";
        return namedParameterJdbcTemplate.update(sql, Map.of("id", id));
    }
}

@Service
public class StationService {

    private static final int NO_ROW_AFFECTED = 0;
    private static final String STATION_DUPLICATED = "이미 존재하는 지하철역입니다. ";
    private static final String STATION_NOT_FOUND = "요청한 지하철 역이 존재하지 않습니다. ";

    ...

    @Transactional
    public Station save(final StationRequest stationRequest) {
        final Station station = new Station(stationRequest.getName());
        try {
            return stationDao.save(station);
        } catch (
        DuplicateKeyException e) {
            throw new IllegalStateException(STATION_DUPLICATED + station);
        }
    }

    ...

    @Transactional
    public void deleteById(final Long id) {
        final int theNumberOfAffectedRow = stationDao.deleteById(id);
        if (theNumberOfAffectedRow == NO_ROW_AFFECTED) {
            throw new IllegalStateException(STATION_NOT_FOUND + "id=" + id);
        }
    }
}

skullkim committed:
feat(Section.java): 다른 색션과 비교해 같은역 존재를 판단하는 기능 완료

skullkim committed:
feat(Section.java):  두 섹션의 거리를 비교하는 기능 완료

skullkim committed:
feat(StationFactory.java): 테스트를 위한 정거장 팩토리 구현 완료

중복되는 정거장 인스턴스의 생성을 막기 위해 정거장 정적 팩토리 구현

skullkim committed:
fix(Section.java): 세션 매치 여부 검사 로직 변경

새로운 섹션 추가를 위해 섹션간 정거장 일치여부를 검사하는 로직을
요구사항에 맞게 수정

BREAKING CAHNGE: match 메서드 로직 변경
as-is:
public class Section {
    ...

    public MatchingResult match(final Section newSection) {
        if (upStation.equals(newSection.upStation) && downStation.equals(newSection.downStation)) {
            return MatchingResult.SAME_SECTION;
        }
        if (upStation.equals(newSection.upStation)) {
            return MatchingResult.SAME_UP_STATION;
        }
        if (downStation.equals(newSection.downStation)) {
            return MatchingResult.SAME_DOWN_STATION;
        }
        return MatchingResult.NO_MATCHED;
    }

    ...
}

to-be:
public class Section {
    ...

    public MatchingResult match(final Section newSection) {
        if (isSameSection(newSection)) {
            return MatchingResult.SAME_SECTION;
        }
        if (canAddToLeft(newSection)) {
            return MatchingResult.ADD_TO_LEFT;
        }
        if (canAddToRight(newSection)) {
            return MatchingResult.ADD_TO_RIGHT;
        }
        return MatchingResult.NO_MATCHED;
    }

    ...
}

skullkim committed:
refactor(Section.java): 섹션 일치 여부 검증 로직 변경

섹션 일치 여부를 검증하는 로직 단순화

BREAKING CHANGE: match 메서드 로직 단순화, 출발역과 도착역에 대한 검증
분리
as-is:
public class Section {
    ...

    public MatchingResult match(final Section newSection) {
        ...
        if (canAddToLeft(newSection)) {
            return MatchingResult.ADD_TO_LEFT;
        }
        if (canAddToRight(newSection)) {
            return MatchingResult.ADD_TO_RIGHT;
        }
        return MatchingResult.NO_MATCHED;
    }

    ...

    private boolean canAddToLeft(final Section newSection) {
        return upStation.equals(newSection.downStation)
                ^ downStation.equals(newSection.downStation);
    }

    private boolean canAddToRight(final Section newSection) {
        return upStation.equals(newSection.upStation)
                ^ downStation.equals(newSection.upStation);
    }

    public boolean isDistanceLongerThan(final Section newSection) {
        return this.distance > newSection.distance;
    }
}

to-be:
public class Section {
    ...

    public MatchingResult match(final Section newSection) {
        ...
        if (this.downStation.equals(newSection.downStation)) {
            return MatchingResult.ADD_TO_LEFT;
        }
        if (this.upStation.equals(newSection.upStation)) {
            return MatchingResult.ADD_TO_RIGHT;
        }
        return MatchingResult.NO_MATCHED;
    }

    public MatchingResult matchStartStation(final Section newSection) {
        if (this.upStation.equals(newSection.downStation)) {
            return MatchingResult.ADD_TO_LEFT;
        }
        return MatchingResult.NO_MATCHED;
    }

    public MatchingResult matchEndStation(final Section newSection) {
        if (this.downStation.equals(newSection.upStation)) {
            return MatchingResult.ADD_TO_RIGHT;
        }
        return MatchingResult.NO_MATCHED;
    }

    ...
}

skullkim committed:
feat(Sections.java): 새로운 섹션을 추가하는 기능 완료

새로운 색션이 들어갈 위치를 판단해 섹션을 추가하는 기능 완료

skullkim committed:
refactor(Section.java): getter 삭제

사용하지 않는 getter 삭제

BREAKING CHANGE: getUpStation(), getDownStation() 삭제
as-is:
public class Section {
    ...

    public Station getUpStation() {
        return upStation;
    }

    public Station getDownStation() {
        return downStation;
    }

    ...
}

to-be:
public class Section {
    ...
}

skullkim committed:
refactor(SectionsTest.java): 테스트 코드 통합

두 개로 나눠져 있던 테스트를 하나의 ParameterizedTest로 통합

BREAKING CHANGE: 섹션 추가 테스트를 하나로 통합
as-is:
@DisplayName("Sections 는")
class SectionsTest {

    @DisplayName("추가할 섹션의 위치를 찾고 섹션을 추가해야 한다.")
    @Test
    void add_Section() {
        final Sections sections = new Sections(new ArrayList<>());
        sections.addIfPossible(SectionFactory.from("ab3"));
        sections.addIfPossible(SectionFactory.from("ac3"));
        assertThat(sections).extracting("value").isEqualTo(List.of(SectionFactory.from("ac3"),
                SectionFactory.from("cb3")));
    }

    @DisplayName("추가할 섹션의 위치를 찾고 섹션을 추가해야 한다.")
    @Test
    void add_Section2() {
        final Sections sections = new Sections(new ArrayList<>());
        sections.addIfPossible(SectionFactory.from("ab3"));
        sections.addIfPossible(SectionFactory.from("cb3"));
        assertThat(sections).extracting("value").isEqualTo(List.of(
                SectionFactory.from("ac3"),SectionFactory.from("cb3")));
    }
}

to-be:
@DisplayName("Sections 는")
class SectionsTest {

    @DisplayName("추가할 섹션의 위치를 찾고 섹션을 추가해야 한다.")
    @ParameterizedTest(name = "{index} {displayName} section={0} expectedResult={1}")
    @MethodSource("provideAddSectionSource")
    void add_Section(final Section section, final List<Section> expectedResult) {
        final Sections sections = new Sections(new ArrayList<>());
        sections.addIfPossible(SectionFactory.from("ab3"));
        sections.addIfPossible(section);
        assertThat(sections).extracting("value").isEqualTo(expectedResult);
    }

    private static Stream<Arguments> provideAddSectionSource() {
        return Stream.of(
                Arguments.of(SectionFactory.from("ac3"), List.of(SectionFactory.from("ac3"),
                        SectionFactory.from("cb3"))),
                Arguments.of(SectionFactory.from ("cb3"), List.of(
                        SectionFactory.from("ac3"),SectionFactory.from("cb3")))
        );
    }
}

skullkim committed:
feat(Sections.java): 섹션 삭제 기능 완료

skullkim committed:
feat(Sections.java): 거리 검증 추가

구간을 추가하려 할 때 구간 끼리의 거리를 검증하는 로직 추가

skullkim committed:
chore(dto/): 하위 패키지 생성

관리를 용이하게 하기 위해 dto의 하위 패키지 생성

BREAKING CHANGE: request/, response/ 패키지 생성

skullkim committed:
feat(SectionDao.java): 구간 DAO완료

구간에 대한 테이블에 접근하는 DAO 완료

skullkim committed:
feat(StationRepository.java): 역에 관련된 레파지토리 생성 완료

skullkim committed:
feat(StationDao.java): 특정역 검색 기능 완료

특정역을 아이디로 검색해 가져오는 기능 완료

skullkim committed:
feat(LineRepository.java): 라인 추가 기능 완료

skullkim committed:
feat(LineRepository.java): 특정 노선을 찾는 기능 완료

skullkim committed:
feat(LineRepository.java): 모든 호선을 조회하는 기능 완료

skullkim committed:
fix(test/): 테스트 수정

구현 코드 변경으로 인해 꺠지는 테스트 수정

skullkim committed:
feat(SectionService.java): 섹션 서비스 완료

skullkim committed:
refactor(SectionController.java): 구간에 대한 컨트롤러 생성

hongbin-dev commented on src/main/java/wooteco/subway/domain/MatchingResult.java:
```diff
@@ -0,0 +1,9 @@
+    SAME_SECTION,
+    NO_MATCHED,
+    POSSIBLE_TO_DELETE
+}
```
어떤 액션을 취할지 Result를 뽑아냈군요 😮
이 객체에 좀 더 책임이 있으면 좋겠어요. 또 ADD액션과 DELETE ACTION을 나눠주면 좋을 것 같아요.

skullkim replied:
수정했습니다.

hongbin-dev commented on src/main/java/wooteco/subway/domain/Section.java:
```diff
@@ -0,0 +1,139 @@
+            return MatchingResult.ADD_TO_RIGHT;
+        }
+        return MatchingResult.NO_MATCHED;
+    }
```
MatchResult의 책임으로 옮기는건 어떨까요?

skullkim replied:
수정했습니다.

hongbin-dev commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,151 @@
+    private static final String CANNOT_ADD_SECTION = "섹션을 추가할 수 없습니다.";
+    private static final String CANNOT_DELETE_SECTION = "역을 삭제할 수 없습니다.";
+
+    private List<Section> value;
```
가변으로 둔 이유가 있을까요?

skullkim replied:
삭제와 추가시 가변이 더 용이하다 판단해 가변으로 두었습니다. addIfPossible과 deleteIfPossible의 로직을 보시면 가변이여야 가능한 로직들이 포함되 있습니다.

hongbin-dev commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,151 @@
+                return;
+            }
+        }
+        throw new IllegalArgumentException(CANNOT_ADD_SECTION + " " + newSection);
```
예외를 먼저 파악할 순 없을까요?

또 더하는 비즈니스 순서에 의존적이라 어떤 이유로 실패하는지 이유를 알 수 없을 것 같아요.

skullkim replied:
주신 피드백이 잘 이해가 안되는데 조금만 더 설명해 주실 수 있나요? 피드백을 주신 부분에는 여러 분기 처리가 존재하긴 하지만 이들은 값을 추가하는 로직이 달라서 분기처리를 한거에요. 따라서 addIfPossible() 메서드는 오직 새로운 섹션을 추가하지 못할 때에만 예외를 던져요. 그런데 어떤 이유에서 "어떤 이유로 실패하는지 이유를 알 수 없다"라고 하신건가요?

"더하는 비즈니스 순서에 의존적"이라는 것이 혹시 현재 메서드 내의 조건문과 반복문의 순서를 바꾸면 제대로 동작을 하지 않는다는 의미인가요? 이 의도가 맞다면, 이 메서드는 그렇지 않아요. 위에서 언급했듯이 단순히 섹션을 추가하는 로직이 경우에 따라 달라서 분기처리를 한것이라 순서를 바꾸어도 상관없어요.

예외를 먼저 파악할려면 현재 메서드 내에 존재하는 조건문의 조건들을  한곳에 모아서 하나라도 참인지를 판단하고, 모두 false면 예외를 던져야죠. 만약 하나라도 참이면 같은 조건들을 한번 더 검사하면서 어디에 어떻게 추가할지 분기처리를 해야 해요. 즉, 같은 연산을 반복하게 되는데 이 방식이 더 효율적인가요?

hongbin-dev replied:
제가 말을 잘못썼군요 

`또 더하는 비즈니스 순서에 의존적이라 어떤 이유로 실패하는지 이유를 알 수 없을 것 같아요.` 라기보단 `위 비즈니스에 의존적인 코드라서 어떤 이유로 실패하는지 파악하기 어렵다` 입니다. 죄송해요 😅

```
if (value.isEmpty() || isNewDownSection(newSection)) {
            value.add(newSection);
            return;
        }
        if (isNewUpSection(newSection)) {
            value = addFirst(newSection);
            return;
        }
...
```

이런 코드가 모두 실행되고, 마지막라인까지 살아남은코드가 예외가 발생할 것 같은데요. 그게 if-else문의 else와 비슷하다고 생각했어요. else를 읽을려면 if, elseif의 모든 코드를 읽고나서 파악할 수 있지 않을까요?

피드백은 빠르게 실행되어야한다고 생각하는데요. 예외를 먼저 발생시켜서 읽기 좋게만들 수 없을까? 의 피드백이었습니다.

hongbin-dev commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,151 @@
+                && section.isDistanceLongerThan(newSection);
+    }
+
+    private void add(final Section section, final Section newSection, final MatchingResult result) {
```
메서드 네이밍을 더 신경썼으면 좋겠습니다! 어떨 때 더하는 메서드인가요?

skullkim replied:
수정했습니다.

hongbin-dev commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,151 @@
+        leftSection.remove(leftSection.size() - 1);
+        leftSection.add(seperatedSection.get(0));
+        leftSection.addAll(rightSection);
+        value = new ArrayList<>(leftSection);
```
개행이 없어서 비즈니스를 읽기가 조금 어렵네요 ㅜㅜ

skullkim replied:
수정했어요.

hongbin-dev commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,151 @@
+            }
+        }
+        throw new IllegalArgumentException(CANNOT_DELETE_SECTION + " " + target);
+    }
```
삭제할 수 없으면, 빠르게 예외를 내서 피드백받는게 좋을 것 같습니다.

skullkim replied:
수정했습니다.

hongbin-dev commented on src/main/java/wooteco/subway/dto/request/LineRequest.java:
```diff
@@ -0,0 +1,52 @@
+    private int distance;
+
+    public LineRequest() {
+    }
```
일반클라이언트가 생성할 수 없도록 접근지정자를 막는게 좋을 것 같습니다!

skullkim replied:
수정했습니다.

hongbin-dev commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -1,49 +1,63 @@
-    public Line save(final LineRequest lineRequest) {
-        final Line line = new Line(lineRequest.getName(), lineRequest.getColor());
-        return lineDao.save(line);
+//    @Transactional
```
트랜잭션을 지운 이유가 무엇인가요? 또 불필요한 주석이라면 다 지워주시면 좋을 것 같습니다

skullkim replied:
수정했어요.

hongbin-dev commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -1,49 +1,63 @@
+    private final CommonLineDao lineDao;
+    private final LineRepository lineRepository;
+
+    public LineService(final CommonLineDao lineDao, final LineRepository lineRepository) {
```
dao와 repository는 무엇이 다른건가요? 패러다임이 겹치는 것 같은데요. 둘 다 사용하는 이유가 궁금합니다.

skullkim replied:
service는 도메인과 디비에서 가져온 데이터에 대한 상호작용을 관여하고, dao는 오직 데이터 접근에 대해서만 관여하고, repository는 2개 이상의 디비에 대한 접근과 가져온 데이터에 대한 일부 처리를 담당해야 된다고 생각합니다. 그럼에도 LineService를 저렇게 구현한 이유는 시간이 없어서 못나눠서 그래요..😭 . 분리했습니다

hongbin-dev replied:
`repository는 2개 이상의 디비에 대한 접근` 이 부분은 서비스의 책임이 아닐까요?
레이어가 많아지는 것, 영속성레이어를 다루는걸로 알려진 repository가 다른 역할을 하고 있어서 파악이 어려운 부분이 있네요

hongbin-dev commented on src/main/java/wooteco/subway/service/SectionService.java:
```diff
@@ -0,0 +1,71 @@
+import wooteco.subway.repository.StationRepository;
+import wooteco.subway.service.dto.SectionDto;
+
+@Service
+public class SectionService {
```
SectionService에 트랜잭션을 보장해주면 좋을 것 같습니다~

skullkim replied:
SectionRepository를 생성하고 repository에 트랜잭션을 보장하게 했습니다.

hongbin-dev commented on src/main/java/wooteco/subway/service/SectionService.java:
```diff
@@ -0,0 +1,71 @@
+        this.stationRepository = stationRepository;
+    }
+
+    public void save(final Long lineId, final SectionRequest request) {
```
여기도 개행이 없으니 가독성이 떨어지는 것 같습니다! 비즈니스흐름대로 개행을 넣어보는것도 가독성이 좋아질 수 있어요.

skullkim replied:
수정하고 보지 메서드도 길어지면 개행을 넣는 것이 가독성에 좋은거 같네요. 수정했습니다.

hongbin-dev commented on src/main/java/wooteco/subway/service/SectionService.java:
```diff
@@ -0,0 +1,71 @@
+        addSections(lineId, addSections);
+    }
+
+    private void addSections(final Long lineId, final List<Section> sections) {
+        for (Section section : sections) {
+            sectionDao.save(lineId, SectionDto.from(section));
+        }
+    }
+
+    private void deleteSections(final Long lineId, final List<Section> sections) {
+        for (Section section : sections) {
+            sectionDao.deleteById(lineId, section.getUpStationId());
+        }
+    }
```
N번의 저장과 삭제가 일어나지않을까요?
현재시점에 N번의 커넥션과 쿼리가 나갈 것 같은데, 다른방법이 있을까요?

skullkim replied:
디비에 저장된 레코드 중에 변경된 부분만 addSections와 deleteSections로 추가하고, 삭제하는데 더 좋은 방법이 있나요? 이 두 메서드는 사용하기 전에 중복되는 부분을 삭제하는 로직을 거친 후 사용되요. 예를 돌이 역을 삭제하는 메서드 내에서는 이 두 메서드가 다음과 같이 사용되요.
```java
public void delete(final Long lineId, final Long stationId) {
        final Line line = lineRepository.findById(lineId);
        final Station target = stationRepository.findById(stationId);
        // 기존 섹션들을 불러온다.
        final List<Section> previousSections = new ArrayList<>(line.getSections());
       // 삭제하길 원하는 역을 지운다
        line.deleteSection(target);
        // 기존 섹션(previousSections)과 새로운 섹션들을 비교해 새로 추가된 섹션들을
       // 가져온다.
        final List<Section> addSections = line.getAddSections(previousSections);
       // 기존 섹션(previousSections)과 새로운 섹션들을 비교해 삭제된 섹션들을 
       // 가져온다
        final List<Section> deletedSections = line.getDeletedSections(previousSections);
       // 위에서 가져온 삭제된 섹션들을 디비에서 지운다
        deleteSections(lineId, deletedSections);
       // 위에서 가져온 추가된 섹션들을 디비에 추가한다.
        addSections(lineId, addSections);
    }
```
따라서 중복을 없애고, 반드시 필요한 섹션들에 한해서만 쿼리를 날리고 있어요. 지우고 삭제하는게 아니라 update도 고려해봤어요. 하지만 어떤 역을 지우냐에 따라 한 section의 upStation 또는 downStation만 바뀌지 않고 새로운 section이 생기는 경우, 기존 section이 아예 없어지는 경우도 있어서 이런 로직을 고안했어요.
앨런이 생각하는 더 좋은 방법이 있을까요?

hongbin-dev replied:
스컬 상세한 설명 감사합니다! 스컬이 작성해준 서비스코드 잘 이해됐습니다.
이야기 드리고싶은 부분은 쿼리레벨이었어요. 슈도코드로 설명드릴께요!

```java
// AS-IS for문으로 1번씩 쿼리가 N번 발생
for (Section section : sections) {
            sectionDao.deleteById(lineId, section.getUpStationId());
        }

// TO-BE 1번의 쿼리로 모두 삭제
List<Long> ids = sections.stream...
sectionDao.deleteByIds(ids)
```

hongbin-dev commented on src/main/java/wooteco/subway/service/StationService.java:
```diff
@@ -1,39 +1,67 @@
+        stationRepository.deleteById(id);
     }
+
+    //    private static final int NO_ROW_AFFECTED = 0;
+//    private static final String STATION_DUPLICATED = "이미 존재하는 지하철역입니다. ";
+//    private static final String STATION_NOT_FOUND = "요청한 지하철 역이 존재하지 않습니다. ";
+//
+//    private final CommonStationDao stationDao;
+//
+//    public StationService(final CommonStationDao stationDao) {
+//        this.stationDao = stationDao;
+//    }
+//
+//    @Transactional
+//    public Station save(final StationRequest stationRequest) {
+//        final Station station = new Station(stationRequest.getName());
+//        try {
+//            return stationDao.save(station);
+//        } catch (
+//        DuplicateKeyException e) {
+//            throw new IllegalStateException(STATION_DUPLICATED + station);
+//        }
+//    }
+//
+//
+//    @Transactional(readOnly = true)
+//    public List<Station> findAll() {
+//        return stationDao.findAll();
+//    }
+//
+//
+//    @Transactional
+//    public void deleteById(final Long id) {
+//        final int theNumberOfAffectedRow = stationDao.deleteById(id);
+//        if (theNumberOfAffectedRow == NO_ROW_AFFECTED) {
+//            throw new IllegalStateException(STATION_NOT_FOUND + "id=" + id);
+//        }
+//    }
```
이쪽도 사용하지않으면 모두 제거해주세요~

skullkim replied:
수정했어요.

hongbin-dev commented on src/test/java/wooteco/subway/domain/SectionFactory.java:
```diff
@@ -0,0 +1,28 @@
+
+    static {
+        cache = new HashMap<>();
+        cache.put("ab3", new Section(StationFactory.from("a"), StationFactory.from("b"), 3));
+        cache.put("ac3", new Section(StationFactory.from("a"), StationFactory.from("c"), 3));
+        cache.put("cb3", new Section(StationFactory.from("c"), StationFactory.from("b"), 3));
+        cache.put("cd3", new Section(StationFactory.from ("c"), StationFactory.from("d"), 3));
+        cache.put("ca3", new Section(StationFactory.from("c"), StationFactory.from("a"), 3));
+        cache.put("ba3", new Section(StationFactory.from("b"), StationFactory.from("a"), 3));
+        cache.put("bc3", new Section(StationFactory.from("b"), StationFactory.from("c"), 3));
+        cache.put("ac2", new Section(StationFactory.from("a"), StationFactory.from("c"), 2));
+        cache.put("ac1", new Section(StationFactory.from("a"), StationFactory.from("c"), 1));
+        cache.put("cb1", new Section(StationFactory.from("c"), StationFactory.from("b"), 1));
+        cache.put("cb2", new Section(StationFactory.from("c"), StationFactory.from("b"), 2));
```
ab3 이런 값은 public 상수로 관리해도 좋겠군요~

skullkim replied:
중복된 부분이 많네요, 수정하겠습니다.

hongbin-dev replied:
중복제거의 이유도 있지만, 외부에서 사용할 때 `SectionFactory.AB3` 이런식으로 값을 쓸 수 있겠군요~

hongbin-dev commented:
스컬! 테스트가 많이 깨지는게 확인 한번해주세요~

hongbin-dev reviewed (changes requested):
안녕하세요 스컬!
3단계 미션 많이 어려운데 잘 구현해주셨네요.
코멘트와 질문 남겼으니 확인해주세요~

skullkim committed:
fix(SectionsTest.java): 실패하는 테스트 수정

잘못된 given으로 인해 실패하는 테스트를 통과하게 수정

BREKAING CHANGE: 테스트에 사용되는 주어지는 Section 인스턴스 수정
as=is:
@DisplayName("Sections 는")
class SectionsTest {
    ...

    @DisplayName("지울 역을 찾고 역을 지워야 한다.")
    @ParameterizedTest(name = "{index} {displayName} section={0} expectedResult={1}")
    @MethodSource("provideDeleteSectionSource")
    void delete_Section(final Section section, final Station targetStation, final List<Section> expectedResult) {
        ...
    }

    private static Stream<Arguments> provideDeleteSectionSource() {
        return Stream.of(
                Arguments.of(SectionFactory.from("ac3"), StationFactory.from("c"),
                        List.of(SectionFactory.from("ab3"))),
                ...
            );
    }

    ...
}

to-be:
@DisplayName("Sections 는")
class SectionsTest {
    ...

    @DisplayName("지울 역을 찾고 역을 지워야 한다.")
    @ParameterizedTest(name = "{index} {displayName} section={0} expectedResult={1}")
    @MethodSource("provideDeleteSectionSource")
    void delete_Section(final Section section, final Station targetStation, final List<Section> expectedResult) {
        ...
    }

    private static Stream<Arguments> provideDeleteSectionSource() {
        return Stream.of(
                Arguments.of(SectionFactory.from("ac3"), StationFactory.from("c"),
                        List.of(SectionFactory.from("ab3"))),
                ...
            );
    }

    ...

}

skullkim committed:
fix(LineAcceptanceTest.java): 깨지는 테스트 수정

요청에 필요한 데이터가 바뀜에 따라 꺠지는 테스트를 통과되게 수정

BREAKING CHANGE: 라인 생성 전 필요한 역을 생성하는 기능 추가
as-is:
@DisplayName("노선 관련 기능")
public class LineAcceptanceTest extends AcceptanceTest {
    ...

     private ExtractableResponse<Response> createLine(final String name, final String color) {
        final Map<String, String> params = new HashMap<>();
        params.put("name", name);
        params.put("color", color);
        return RestAssured.given().log().all()
            ...
        }

    ...
}

to-be:
@DisplayName("노선 관련 기능")
public class LineAcceptanceTest extends AcceptanceTest {
    ...

    private ExtractableResponse<Response> createLineAfterCreatingStation(final String name, final String color,
                                                                         final String upStationName,
                                                                         final String downStationName,
                                                                         final int distance) {
        final ExtractableResponse<Response> stationResponse1 = createStation(upStationName);
        final ExtractableResponse<Response> stationResponse2 = createStation(downStationName);
        final Long upStationId = stationResponse1.jsonPath().getObject(".", StationResponse.class).getId();
        final Long downStationId = stationResponse2.jsonPath().getObject(".", StationResponse.class).getId();
        final Map<String, Object> params = new HashMap<>();
        params.put("name", name);
        params.put("color", color);
        params.put("upStationId", upStationId);
        params.put("downStationId", downStationId);
        params.put("distance", distance);
        return RestAssured.given().log().all()
                .body(params)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/lines")
                .then().log().all()
                .extract();
    }

    ...

    private ExtractableResponse<Response> createStation(final String name) {
        final Map<String, String> params = new HashMap<>();
        params.put("name", name);
        return RestAssured.given().log().all()
                .body(params)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/stations")
                .then().log().all()
                .extract();
    }

}

skullkim committed:
chore(StationService.java): 필요 없는 주석 삭제

skullkim committed:
chore(LineService.java): 필요없는 주석 삭제

skullkim committed:
refactor(MatchingResult.java): 책임 추가, 책임 분리

MathingResult는 섹션 추가에 대한 결과만을 판별하는 책임을 갖게 된다.
기존 섹션에 대한 객체가 가지고 있던, 섹션 추가 가능 여부를 판단하는
책임을 MatchingResult로 전가

BREAKING CHANGE: Section 객체가 가지는 책임을 MatchingResult로 전가
as-is:
public enum MatchingResult {
    ADD_TO_RIGHT,
    ADD_TO_LEFT,
    SAME_SECTION,
    NO_MATCHED,
    POSSIBLE_TO_DELETE
}

public class Section {

    ...

    public MatchingResult match(final Section newSection) {
        if (isSameSection(newSection)) {
            return MatchingResult.SAME_SECTION;
        }
        if (this.downStation.equals(newSection.downStation)) {
            return MatchingResult.ADD_TO_LEFT;
        }
        if (this.upStation.equals(newSection.upStation)) {
            return MatchingResult.ADD_TO_RIGHT;
        }
        return MatchingResult.NO_MATCHED;
    }

    public MatchingResult matchStartStation(final Section newSection) {
        if (this.upStation.equals(newSection.downStation)) {
            return MatchingResult.ADD_TO_LEFT;
        }
        return MatchingResult.NO_MATCHED;
    }

    public MatchingResult matchEndStation(final Section newSection) {
        if (this.downStation.equals(newSection.upStation)) {
            return MatchingResult.ADD_TO_RIGHT;
        }
        return MatchingResult.NO_MATCHED;
    }

    ...
}

to-be:
public enum MatchingResult {
    ADD_TO_RIGHT,
    ADD_TO_LEFT,
    SAME_SECTION,
    NO_MATCHED
    ;

    public static MatchingResult matchMiddleStation(final Section exSection, final Section newSection) {
        if (exSection.isSameSection(newSection)) {
            return SAME_SECTION;
        }
        if (exSection.isSameDownStation(newSection)) {
            return ADD_TO_LEFT;
        }
        if (exSection.isSameUpStation(newSection)) {
            return ADD_TO_RIGHT;
        }
        return NO_MATCHED;
    }

    public static MatchingResult matchStartStation(final Section exSection, final Section newSection) {
        if (exSection.isUpStationSameAsDownStation(newSection)) {
            return MatchingResult.ADD_TO_LEFT;
        }
        return MatchingResult.NO_MATCHED;
    }

    public static MatchingResult matchEndStation(final Section exSection, final Section newSection) {
        if (exSection.isDownStationSameAsUpStation(newSection)) {
            return MatchingResult.ADD_TO_RIGHT;
        }
        return MatchingResult.NO_MATCHED;
    }
}

public class Section {
    ...

    public boolean isSameSection(final Section newSection) {
        return isSameUpStation(newSection) && isSameDownStation(newSection);
    }

    public boolean isSameDownStation(final Section newSection) {
        return downStation.equals(newSection.downStation);
    }

    public boolean isSameUpStation(final Section newSection) {
        return upStation.equals(newSection.upStation);
    }

    public boolean isUpStationSameAsDownStation(final Section newSection) {
        return upStation.equals(newSection.downStation);
    }

    public boolean isDownStationSameAsUpStation(final Section newSection) {
        return downStation.equals(newSection.upStation);
    }

    ...

}

skullkim committed:
refactor(MatchingResult): 객체 이름 변경

이름을 명확히 하기 위해 이름 변경

BREAKING CHANGE: MatchingResult 객체 이름 변경
as-is:
src/main/java/wooteco/subway/domain/MatchingResult.java

to-be:
src/main/java/wooteco/subway/domain/AddMatchingResult.java

skullkim committed:
refactor(DeleteMatchingResult.java): 객체의 책임 이동

섹션에 대한 객체가 가지고 있던 책임을 DeleteMatchingResult로 이동

BREKAING CHANGE: 섹션 삭제 가능 여부를 판단하는 로직 이동
as-is:
public enum DeleteMatchingResult {
    POSSIBLE_TO_DELETE,
    NO_MATCHED,
}

public class Section {
    ...

    public DeleteMatchingResult matchStation(final Station target) {
        if (this.downStation.isSameName(target)) {
            return DeleteMatchingResult.POSSIBLE_TO_DELETE;
        }
        return DeleteMatchingResult.NO_MATCHED;
    }

    public DeleteMatchingResult matchWithLastUpStation(final Station target) {
        if (this.upStation.isSameName(target)) {
            return DeleteMatchingResult.POSSIBLE_TO_DELETE;
        }
        return DeleteMatchingResult.NO_MATCHED;
    }

    public DeleteMatchingResult matchWithLastDownStation(final Station taret) {
        if (this.downStation.isSameName(taret)) {
            return DeleteMatchingResult.POSSIBLE_TO_DELETE;
        }
        return DeleteMatchingResult.NO_MATCHED;
    }

    ...
}

to-be:
public enum DeleteMatchingResult {

    POSSIBLE_TO_DELETE,
    NO_MATCHED
    ;

    public static DeleteMatchingResult matchStation(final Section section, final Station target) {
        if (section.isSameWithDownStation(target)) {
            return POSSIBLE_TO_DELETE;
        }
        return NO_MATCHED;
    }

    public static DeleteMatchingResult matchWithLastUpStation(final Section section, final Station target) {
        if (section.isSameWithUpStation(target)) {
            return POSSIBLE_TO_DELETE;
        }
        return NO_MATCHED;
    }

    public static DeleteMatchingResult matchWithLastDownStation(final Section section, final Station taret) {
        if (section.isSameWithDownStation(taret)) {
            return POSSIBLE_TO_DELETE;
        }
        return NO_MATCHED;
    }
}

public class Section {
    ...

        public boolean isSameWithDownStation(final Station target) {
        return this.downStation.isSameName(target);
    }

    public boolean isSameWithUpStation(final Station target) {
        return this.upStation.isSameName(target);
    }

    ...
}

skullkim committed:
refactor(Sections.java): 메서드 이름 변경

메서드의 의도를 더 명확히 하기 위해 메서드 이름 변경

BREAKING CHANGE: add() 메서드 이름 변경
as-is:
public class Sections {
    ...

    private void add(final Section section, final Section newSection, final AddMatchingResult result) {
        ...
    }

    ...
}

to-be:
public class Sections {
    ...

    private void addSection(final Section section, final Section newSection, final AddMatchingResult result) {
        ...
    }

    ...
}

skullkim committed:
chore(Sections.java): 메서드 사이에 공백 추가

길이가 긴 메서드의 가독성 향상을 위해 메서드 사이에 공백 추가

skullkim committed:
refactor(Sections.java): 역을 지울수 없을 떄 에러를 반환하게 변경

역을 지울 수 없을 때 그냥 종료를 하는 대신 예외를 던지는 방식으로 변경

BREAKING CHANGE: canDelete()가 false일 떄 처리 로직 변경
as-is:
public class Sections {
    ...

    public void deleteIfPossible(final Station target) {
        if (!canDelete()) {
            return;
        }
        ...
    }

    ....
}

to-be:
public class Sections {
    ...

    public void deleteIfPossible(final Station target) {
        if (!canDelete()) {
            throw new IllegalArgumentException(CANNOT_DELETE_SECTION + " " + target);
        }
        ...
    }

    ....
}

skullkim committed:
chore(SectionService.java): 메서드 내에 개행 추가

코드의 가독성을 높이기 위해 메서드에 개행 추가

skullkim committed:
refactor(subway/dto/): 기본 생성자의 접근 제어자 변경

일반 클래스의 접근을 막기 위해 기본 생성자의 접근 제어자 변경

BREAKING CHANGE: 기본 생성자의 접근 제어자를 public 에서 private으로
변경

skullkim committed:
refactor(/test/java/wooteco/subway/domain): 팩토리 키를 상수로 변경

섹션과 역에 대한 팩토리에서 객체를 찾는 키를 스트링에서 상수로 변경

BREAKING CHANGE: SectionFactory, StationFactory의 cache의 key에 상수
적용
as-is:
public class StationFactory {

    private static final Map<String, Station> cache;

    static {
        cache = new HashMap<>();
        cache.put("a", new Station(1L, "a"));
        cache.put("b", new Station(2L, "b"));
        cache.put("c", new Station(3L, "c"));
        cache.put("d", new Station(4L, "d"));
    }

    ...
}

public class SectionFactory {

    private static final Map<String, Section> cache;

    static {
        cache = new HashMap<>();
        cache.put("ab3", new Section(StationFactory.from("a"), StationFactory.from("b"), 3));
        cache.put("ac3", new Section(StationFactory.from("a"), StationFactory.from("c"), 3));
        cache.put("cb3", new Section(StationFactory.from("c"), StationFactory.from("b"), 3));
        cache.put("cd3", new Section(StationFactory.from ("c"), StationFactory.from("d"), 3));
        cache.put("ca3", new Section(StationFactory.from("c"), StationFactory.from("a"), 3));
        cache.put("ba3", new Section(StationFactory.from("b"), StationFactory.from("a"), 3));
        cache.put("bc3", new Section(StationFactory.from("b"), StationFactory.from("c"), 3));
        cache.put("ac2", new Section(StationFactory.from("a"), StationFactory.from("c"), 2));
        cache.put("ac1", new Section(StationFactory.from("a"), StationFactory.from("c"), 1));
        cache.put("cb1", new Section(StationFactory.from("c"), StationFactory.from("b"), 1));
        cache.put("cb2", new Section(StationFactory.from("c"), StationFactory.from("b"), 2));
    }

    ...
}

to-be:
public class StationFactory {

    public static final String A = "a";
    public static final String B = "b";
    public static final String C = "c";
    public static final String D = "d";

    private static final Map<String, Station> cache;

    static {
        cache = new HashMap<>();
        cache.put(A, new Station(1L, A));
        cache.put(B, new Station(2L, B));
        cache.put(C, new Station(3L, C));
        cache.put(D, new Station(4L, D));
    }

    ...
}

public class SectionFactory {

    public static final String AB3 = "ab3";
    public static final String AC3 = "ac3";
    public static final String CB3 = "cb3";
    public static final String CD3 = "cd3";
    public static final String CA3 = "ca3";
    public static final String BA3 = "ba3";
    public static final String BC3 = "bc3";
    public static final String AC2 = "ac2";
    public static final String AC1 = "ac1";
    public static final String CB1 = "cb1";
    public static final String CB2 = "cb2";

    private static final Map<String, Section> cache;

    static {
        cache = new HashMap<>();
        cache.put(AB3, new Section(StationFactory.from(A), StationFactory.from(B), 3));
        cache.put(AC3, new Section(StationFactory.from(A), StationFactory.from(C), 3));
        cache.put(CB3, new Section(StationFactory.from(C), StationFactory.from(B), 3));
        cache.put(CD3, new Section(StationFactory.from (C), StationFactory.from(D), 3));
        cache.put(CA3, new Section(StationFactory.from(C), StationFactory.from(A), 3));
        cache.put(BA3, new Section(StationFactory.from(B), StationFactory.from(A), 3));
        cache.put(BC3, new Section(StationFactory.from(B), StationFactory.from(C), 3));
        cache.put(AC2, new Section(StationFactory.from(A), StationFactory.from(C), 2));
        cache.put(AC1, new Section(StationFactory.from(A), StationFactory.from(C), 1));
        cache.put(CB1, new Section(StationFactory.from(C), StationFactory.from(B), 1));
        cache.put(CB2, new Section(StationFactory.from(C), StationFactory.from(B), 2));
    }

    ...
}

skullkim committed:
refactor(SectionService.java): 책임 분리

서비스가 가지고 있던 일부 책임을 레파지토리로 분리

BREAKING CHANGE: SectionService의 일부 책임을 SectionRepository로 분리
as-is:
@Service
public class SectionService {

    private final SectionDao sectionDao;
    private final LineRepository lineRepository;
    private final StationRepository stationRepository;

    public SectionService(final SectionDao sectionDao,
                          final LineRepository lineRepository,
                          final StationRepository stationRepository) {
        this.sectionDao = sectionDao;
        this.lineRepository = lineRepository;
        this.stationRepository = stationRepository;
    }

    public void save(final Long lineId, final SectionRequest request) {
        final Line line = lineRepository.findById(lineId);
        final List<Section> previousSections = new ArrayList<>(line.getSections());
        final Section newSection = makeSection(request);

        ....

        deleteSections(lineId, deletedSections);
        addSections(lineId, addSections);
    }

    ...

    public void delete(final Long lineId, final Long stationId) {
        final Line line = lineRepository.findById(lineId);
        final Station target = stationRepository.findById(stationId);
        final List<Section> previousSections = new ArrayList<>(line.getSections());

        ....

        deleteSections(lineId, deletedSections);
        addSections(lineId, addSections);
    }

    private void addSections(final Long lineId, final List<Section> sections) {
        for (Section section : sections) {
            sectionDao.save(lineId, SectionDto.from(section));
        }
    }

    private void deleteSections(final Long lineId, final List<Section> sections) {
        for (Section section : sections) {
            sectionDao.deleteById(lineId, section.getUpStationId());
        }
    }

}

to-be:
@Service
public class SectionService {

    private final SectionRepository sectionRepository;

    public SectionService(final SectionRepository sectionRepository) {
        this.sectionRepository = sectionRepository;
    }

    public void save(final Long lineId, final SectionRequest request) {
        final Line line = sectionRepository.findLineById(lineId);
        final List<Section> previousSections = new ArrayList<>(line.getSections());
        final Section newSection = makeSection(request);

        ....

        sectionRepository.deleteSections(lineId, deletedSections);
        sectionRepository.addSections(lineId, addSections);
    }

    ...

    public void delete(final Long lineId, final Long stationId) {
        final Line line = sectionRepository.findLineById(lineId);
        final Station target = sectionRepository.findStationById(stationId);
        final List<Section> previousSections = new ArrayList<>(line.getSections());

        ....

        sectionRepository.deleteSections(lineId, deletedSections);

        sectionRepository.deleteSections(lineId, deletedSections);
        sectionRepository.addSections(lineId, addSections);
    }
}

@Repository
public class SectionRepository {

    private final CommonStationDao stationDao;
    private final CommonSectionDao sectionDao;
    private final CommonLineDao lineDao;

    public SectionRepository(final CommonStationDao stationDao, final CommonSectionDao sectionDao, final CommonLineDao lineDao) {
        this.stationDao = stationDao;
        this.sectionDao = sectionDao;
        this.lineDao = lineDao;
    }

    @Transactional(readOnly = true)
    public Line findLineById(final Long lineId) {
        final List<SectionDto> sectionDtos = sectionDao.findAllByLineId(lineId);
        final Line line = lineDao.findById(lineId);
        return addSectionsToLine(line, sectionDtos);
    }

    private Line addSectionsToLine(final Line line, final List<SectionDto> sectionDtos) {
        final List<SectionDto> sortedSectionDtos = sortSectionDto(sectionDtos, line.getUpStationId());
        for (SectionDto sectionDto : sortedSectionDtos) {
            final Station upStation = stationDao.findById(sectionDto.getUpStationId());
            final Station downStation = stationDao.findById(sectionDto.getDownStationId());
            line.addSection(new Section(upStation, downStation, sectionDto.getDistance()));
        }
        return line;
    }

    private List<SectionDto> sortSectionDto(List<SectionDto> sectionDtos, final Long upStationId) {
        final List<SectionDto> sortedSectionDtos = new ArrayList<>();
        Long sectionId = upStationId;
        while (!sectionDtos.isEmpty()) {
            final SectionDto sectionDto = findSectionDtoById(sectionDtos, sectionId);
            sortedSectionDtos.add(sectionDto);
            sectionId = sectionDto.getDownStationId();
            sectionDtos = remove(sectionDtos, sectionDto);
        }
        return sortedSectionDtos;
    }

    private SectionDto findSectionDtoById(final List<SectionDto> sectionDtos, final Long id) {
        return sectionDtos.stream()
                .filter(sectionDto -> sectionDto.getUpStationId().equals(id))
                .findFirst()
                .orElseThrow(IllegalArgumentException::new);
    }

    private List<SectionDto> remove(final List<SectionDto> sectionDtos, final SectionDto target) {
        return sectionDtos.stream()
                .filter(sectionDto -> !(sectionDto.getUpStationId().equals(target.getUpStationId())))
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public Station findStationById(final Long stationId) {
        return stationDao.findById(stationId);
    }

    @Transactional
    public void addSections(final Long lineId, final List<Section> sections) {
        for (Section section : sections) {
            sectionDao.save(lineId, SectionDto.from(section));
        }
    }

    @Transactional
    public void deleteSections(final Long lineId, final List<Section> sections) {
        for (Section section : sections) {
            sectionDao.save(lineId, SectionDto.from(section));
        }
    }
}

skullkim committed:
refactor(SectionRepository.java): 중복 코드 삭제

다른 레파지토리와 중복되는 코드 삭제

BREAKING CHANGE: LineDao를 LineRepository로 변경
as-is:
@Repository
public class SectionRepository {

    ...
    private final CommonLineDao lineDao;

    public SectionRepository(final CommonStationDao stationDao, final CommonSectionDao sectionDao, final CommonLineDao lineDao) {
        ...
        this.lineDao = lineDao;
    }

    @Transactional(readOnly = true)
    public Line findLineById(final Long lineId) {
        final List<SectionDto> sectionDtos = sectionDao.findAllByLineId(lineId);
        final Line line = lineDao.findById(lineId);
        return addSectionsToLine(line, sectionDtos);
    }

    private Line addSectionsToLine(final Line line, final List<SectionDto> sectionDtos) {
        final List<SectionDto> sortedSectionDtos = sortSectionDto(sectionDtos, line.getUpStationId());
        for (SectionDto sectionDto : sortedSectionDtos) {
            final Station upStation = stationDao.findById(sectionDto.getUpStationId());
            final Station downStation = stationDao.findById(sectionDto.getDownStationId());
            line.addSection(new Section(upStation, downStation, sectionDto.getDistance()));
        }
        return line;
    }

    private List<SectionDto> sortSectionDto(List<SectionDto> sectionDtos, final Long upStationId) {
        final List<SectionDto> sortedSectionDtos = new ArrayList<>();
        Long sectionId = upStationId;
        while (!sectionDtos.isEmpty()) {
            final SectionDto sectionDto = findSectionDtoById(sectionDtos, sectionId);
            sortedSectionDtos.add(sectionDto);
            sectionId = sectionDto.getDownStationId();
            sectionDtos = remove(sectionDtos, sectionDto);
        }
        return sortedSectionDtos;
    }

    private SectionDto findSectionDtoById(final List<SectionDto> sectionDtos, final Long id) {
        return sectionDtos.stream()
                .filter(sectionDto -> sectionDto.getUpStationId().equals(id))
                .findFirst()
                .orElseThrow(IllegalArgumentException::new);
    }

    private List<SectionDto> remove(final List<SectionDto> sectionDtos, final SectionDto target) {
        return sectionDtos.stream()
                .filter(sectionDto -> !(sectionDto.getUpStationId().equals(target.getUpStationId())))
                .collect(Collectors.toList());
    }

    ...

}

to-be:
@Repository
public class SectionRepository {

    ...
    private final LineRepository lineRepository;

    public SectionRepository(final CommonStationDao stationDao, final CommonSectionDao sectionDao,
                             final LineRepository lineRepository) {
        ...
        this.lineRepository = lineRepository;
    }

    @Transactional(readOnly = true)
    public Line findLineById(final Long lineId) {
        return lineRepository.findById(lineId);
    }

    ...
}

skullkim committed:
chore(src/): 필요 없는 코드 삭제

skullkim committed:
chore(src/): 컨벤션에 맞게 코드 스타일 수정

skullkim committed:
refactor(LineService.java): 메서드 이름 수정

명확하지 않은 메서드 이름 수정

BREAKING CHANGE: 메서드 이름에 대상 명시
as-is:
@Service
public class LineService {
    ...

     public LineResponse save(final LineRequest lineRequest) {
        ...
    }

    ...

    public void update(final Long id, final LineRequest lineRequest) {
        ...
    }

    ...
}

to-be:
@Service
public class LineService {
    ...

     public LineResponse saveLine(final LineRequest lineRequest) {
        ...
    }

    ...

    public void updateLine(final Long id, final LineRequest lineRequest) {
        ...
    }

    ...

}

skullkim committed:
refactor(SectionService.java): 메서드 이름 변경

적용 대상이 명확하지 않은 메서드 이름 변경

BREAKING CHANGE: 섹션 저장, 삭제 메서드 이름 변경
as-is:
@Service
public class SectionService {
    ...

    public void save(final Long lineId, final SectionRequest request) {
        ...
    }

    ...

    public void delete(final Long lineId, final Long stationId) {
        ...
    }
}

to-be:
@Service
public class SectionService {
    ...

    public void saveSection(final Long lineId, final SectionRequest request) {
        ...
    }

    ...

    public void deleteSection(final Long lineId, final Long stationId) {
        ...
    }
}

skullkim committed:
refactor(StationService.java): 메서드 이름 변경

대상이 명확하지 않은 메서드 이름 변경

BREAKING CHANGE: 역을 저장하는 메서드 이름 변경
as-is:
@Service
public class StationService {
    ...

    public Station save(final StationRequest stationRequest) {
        ...
    }

    ...
}

to-be:
@Service
public class StationService {
    ...

    public Station saveStation(final StationRequest stationRequest) {
        ...
    }

    ...
}

hongbin-dev commented on src/test/java/wooteco/subway/domain/factory/SectionFactory.java:
```diff
@@ -0,0 +1,46 @@
+
+public class SectionFactory {
+
+    public static final String AB3 = "ab3";
+    public static final String AC3 = "ac3";
+    public static final String CB3 = "cb3";
+    public static final String CD3 = "cd3";
+    public static final String CA3 = "ca3";
+    public static final String BA3 = "ba3";
+    public static final String BC3 = "bc3";
+    public static final String AC2 = "ac2";
+    public static final String AC1 = "ac1";
+    public static final String CB1 = "cb1";
+    public static final String CB2 = "cb2";
```
피드백 반영 좋네요 👍
테스트에서 사용하는 값들을 테스트픽스쳐라고 부르기도 합니다!

hongbin-dev commented on src/main/java/wooteco/subway/repository/SectionRepository.java:
```diff
@@ -1,24 +1,90 @@
 
-    public void save(final Long lineId, final SectionRequest request) {
+    @Transactional(readOnly = true)
+    public Line findLineById(final Long lineId) {
```
https://github.com/woowacourse/atdd-subway-map/pull/320#discussion_r873925410

DAO레벨에서 join을 통해 값을 한번에 가져오긴 어려울까요?

hongbin-dev commented on src/main/java/wooteco/subway/repository/SectionRepository.java:
```diff
@@ -1,24 +1,90 @@
+            final Station upStation = stationDao.findById(sectionDto.getUpStationId());
+            final Station downStation = stationDao.findById(sectionDto.getDownStationId());
+            line.addSection(new Section(upStation, downStation, sectionDto.getDistance()));
+        }
```
생성자를 통해 완성된 Line을 만들 순 없을까요?
Line이 생성된 이후 section을 더하고 있네요.

hongbin-dev reviewed (approved):
안녕하세요 스컬!
피드백 반영 잘 해주셨군요! 코멘트 관련 된 답변남겼습니다. 
궁금하신점 코멘트나 DM 주세요! 이만 머지하겠습니다!
